### PR TITLE
[MHA][tests] Cpu multi head attention

### PR DESCRIPTION
**Pull request by bghimireamd**
First attempt for cpu implementation of multi-head attention. Only inference is done in this PR.

### CODE REVIEW

File: test/gtest/cpu_multi_head_attention.hpp [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+
+namespace test {
+namespace cpu {
+
+template <typename T>
+auto squareRoot(T value) -> decltype(std::sqrt(value))
+{
+    static_assert(std::is_arithmetic<T>::value, "squareRoot only supports arithmetic types");
+    return std::sqrt(value);
+}
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: Seems it just returns `std::sqrt`, is there any purpose for this?

File: test/gtest/cpu_multi_head_attention.hpp [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+template <typename T>
+void BrodcastParMul(tensor<T> A_mat, tensor<T> B_mat, tensor<T>& ret_value)
+{
+    // std::mutex coutMutex;
+    size_t A_col = A_mat.desc.GetLengths()[3];
+    ret_value.par_for_each([&](size_t b_id, size_t h_id, size_t s_id, size_t d_id) {
+        T sum(0);
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: I would recommend to use `double` as accumulator.

File: test/gtest/cpu_multi_head_attention.hpp [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+}
+
+template <typename T>
+void ParMul(tensor<T> A_mat, tensor<T> B_mat, tensor<T>& ret_value, size_t key_dim = 1)
+{
+    std::mutex coutMutex;
+    size_t A_col = A_mat.desc.GetLengths()[3];
+    ret_value.par_for_each([&](size_t b_id, size_t h_id, size_t s_id, size_t d_id) {
+        T sum(0);
+        std::lock_guard<std::mutex> guard(coutMutex);
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: If `ret_value.par_for_each` access to unique elements of `ret_value` then mutex is not needed, especially before the loop.
Currently it serializes almost all the computation.

File: test/gtest/cpu_multi_head_attention.hpp [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+template <typename T>
+void ParMulAttention(tensor<T> A_mat, tensor<T> B_mat, tensor<T>& ret_value, size_t d_k)
+{
+    std::mutex coutMutex;
+    // sequence length
+    size_t A_col = A_mat.desc.GetLengths()[3];
+    ret_value.par_for_each([&](size_t b_id, size_t h_id, size_t s_id, size_t d_id) {
+        T sum(0);
+        size_t dk_id = d_id % d_k;
+        std::lock_guard<std::mutex> guard(coutMutex);
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: See `void ParMul` comment.

File: test/gtest/cpu_multi_head_attention.hpp [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+    size_t head   = tensor_val.desc.GetLengths()[1];
+    size_t sq_len = tensor_val.desc.GetLengths()[2];
+
+    m_tensor = tensor<T>{batch, head, sq_len, 1};
+
+    for(int b = 0; b < batch; ++b)
+    {
+        for(int h = 0; h < head; ++h)
+        {
+            T max_score_in_row;
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: The value can be assigned to `m_tensor` directly, without this variable.

File: test/gtest/cpu_multi_head_attention.hpp [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+    size_t head   = tensor_val.desc.GetLengths()[1];
+    size_t sq_len = tensor_val.desc.GetLengths()[2];
+
+    zinv_tensor = tensor<T>{batch, head, sq_len, 1};
+
+    for(int b = 0; b < batch; ++b)
+    {
+        for(int h = 0; h < head; ++h)
+        {
+            T sum(0);
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: The value can be assigned to `zinv_tensor` directly, without this variable.

File: test/gtest/cpu_multi_head_attention.hpp [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+        // //   9) Zinv
+        Zinv(q_dot_k_transpose, zinv_tensors);
+        // //  10) Zinv reciprocal
+        ZinvMultiply(q_dot_k_transpose, zinv_tensors);
+
+        // //  11) drop out
+        DropOut(q_dot_k_transpose, cpu_mha_test_case.drop_out_rate);
+
+        // //  11.1) drop out scalse
+        double drop_out_scale = 1.0 / (1.0 - d_k);
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: scale depends on `drop_out_rate`.

**Review by CAHEK7 — CHANGES_REQUESTED** [review]

File: test/gtest/cpu_multi_head_attention.hpp [hunk 8]
```diff
@@ -0,0 +1,513 @@
...
+    trans_mat.par_for_each([&](size_t b_id, size_t sc_id, size_t h_id, size_t sl_id, size_t dk_id) {
+        trans_mat(b_id, sc_id, h_id, sl_id, dk_id) = mat_val(b_id, sc_id, h_id, dk_id, sl_id);
+    });
+}
+
+template <class T>
+void DropOut(tensor<T>& q_dot_k_transpose, const double& drop_out_rate)
+{
+    tensor<T> rand_dis(q_dot_k_transpose.desc.GetLengths());
+    rand_dis.generate(GenData<T>{});
```
Comment (CAHEK7) [inline test/gtest/cpu_multi_head_attention.hpp]: We won't get similar mask tensor on gpu. It would be hardly possible to compare the results.

**Review by CAHEK7 — CHANGES_REQUESTED** [review]
Dropout uses random number generator and if we want to get comparable results, we must use the same generators for GPU and CPU.
Take a look at the dropout tests - they have implemented cpu version of the same generator.

**Comment by bghimireamd** [comment]
> Dropout uses random number generator and if we want to get comparable results, we must use the same generators for GPU and CPU. Take a look at the dropout tests - they have implemented cpu version of the same generator.

Agreed! I will change the code accordingly.

**Comment by CAHEK7** [comment]
I've done some significant code refactoring:

- Separated fp32 and fp8 tests and check f8 version (previously only fp32 version checked).
- Made fp8 test self-sufficient (previously it can't be run without fp32)
- Unbound tests with hardcoded reference data from random number generator
- Used `string_view` instead of strings
- Used proper gtest macros for result verification
- Used float instead of double for some intermediate data (which is supposed to be float)

Also I merged https://github.com/ROCm/MIOpen/pull/2819

**Review by CAHEK7 — APPROVED** [review]
mha_helper may use some more generic variants of operations (like `Dot_4D_4D` and all the transposed variants are identical except the index order), but overall it works.
I did some testing against gpu implementation and it works fine.

It's also a base PR for my numerical fwd-mha testing, so it would be nice to merge it before.

**Comment by CAHEK7** [comment]
Sidenote - this code is a subject to change during integration with GPU implementation.
If somebody has any architecture suggestions - they will be taken into account in the following PRs.

Also this PR is a base for bwd-cpu implementation.

File: test/gtest/mha_helper.hpp [hunk 9]
```diff
@@ -0,0 +1,468 @@
...
+// void SetupMask(tensor<T>& mask)
+// {
+//     mask.par_for_each([&](size_t s_id, size_t p_id) {
+//         // make anything above diagonal inf
+//         if(p_id > s_id)
+//         {
+//             mask(s_id, p_id) = -std::numeric_limits<T>::infinity();
+//         }
+//     });
+// }
```
Comment (binarman) [inline test/gtest/mha_helper.hpp:287]: nitpick: looks redundant or comment/TODO is missing

File: test/gtest/cpu_multi_head_attention.hpp [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    void init()
+    {
+        d_k = cpu_mha_test_case.problem_dimension / cpu_mha_test_case.num_heads;
+
+        // mask = tensor<T>{
+        //     std::vector<int>{cpu_mha_test_case.sequence_length,
+        //     cpu_mha_test_case.sequence_length}};
+        // SetupMask(mask);
```
Comment (binarman) [inline test/gtest/cpu_multi_head_attention.hpp:122]: redundant?

File: test/gtest/cpu_multi_head_attention.hpp [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+        word_position = tensor<InputType>{std::vector<int>{cpu_mha_test_case.batch_size,
+                                                           cpu_mha_test_case.sequence_length,
+                                                           cpu_mha_test_case.problem_dimension}};
+
+        // since Pytorch's Y = X*W_tranpose
+        // cpu_mha_test_case.num_heads, cpu_mha_test_case.problem_dimension, d_k
+        //          need to change the dimension to
+        // cpu_mha_test_case.num_heads, d_k, cpu_mha_test_case.problem_dimension
```
Comment (binarman) [inline test/gtest/cpu_multi_head_attention.hpp:131]: and this

**Review by Vsevolod1983 — APPROVED** [review]

**Comment by CAHEK7** [comment]
@junliume may I ask to merge it.
It prevents https://github.com/ROCm/MIOpen/pull/2875 to be reviewed, since it contains the changes from this PR.

**Review by junliume — APPROVED** [review]

File: test/tensor_util.hpp [hunk 12]
```diff
@@ -122,4 +122,66 @@ void output_tensor_to_bin(const char* fileName, T* data, size_t dataNumItems)
     }
 }
 
+template <typename T>
+void output_tensor_to_screen(const tensor<T>& tensor_val,
```
Comment (amberhassaan) [inline test/tensor_util.hpp:126]: minor: simpler name could be `print_tensor`.

File: test/gtest/mha_helper.hpp [hunk 13]
```diff
@@ -0,0 +1,468 @@
...
+
+float GetF8Scaling(float max_val)
+{
+    constexpr float fp8_E4M3_max = 240.0f;
+
+    return fp8_E4M3_max / max_val;
+}
+
+template <typename T>
+T FindMax4D(const tensor<T>& max_of_tensor)
```
Comment (amberhassaan) [inline test/gtest/mha_helper.hpp:63]: A lot of these helper functions on tensors can benefit from variadic lambda parameters like this: 
```
max_of_tensor.for_each([&] (auto... indices) {
   maxVal = std::max(maxVal, max_of_tensor(indices...));
});
```

**Review by amberhassaan — COMMENTED** [review]
some comments.

**Comment by atamazov** [comment]
@junliume Thanks for renaming this to "[MHA] CPU multi head attention" during merging.

@bghimireamd I highly recommend naming PRs as clearly as possible. For this specific PR I would consider prefixing PR name with `[MHA][tests]` 🤝

**Pull request merged**